src/caffe/solver.cpp

```diff
@@ -227,13 +210,44 @@ void Solver<Dtype>::Solve(const char* resume_file) {
         }
       }
     }
-
     ComputeUpdateValue();
     net_->Update();
+
+    // Save a snapshot if needed.
+    if (param_.snapshot() && (iter_ + 1) % param_.snapshot() == 0) {
+      Snapshot();
+    }
+  }
+}
+
+template <typename Dtype>
+void Solver<Dtype>::PreSolve() {
```

Comment:
I agree that forcing subclassers to call `PreSolve` is awkward. Is there any reason we don't just call `PreSolve` from `Init`? That would simplify the logic and get rid of `initialized_`, as there would be no such thing as an uninitialized solver, and then it makes sense to do the base class initialization right in `Init`. I think I'll try it and update the PR.